src/dune_pkg/solver_env.ml

```diff
@@ -168,9 +198,39 @@ module Variable = struct
           [ `Opam_version, opam_version ]
           ~f:(fun (variable, value) -> Pp.textf "%s = %s" (to_string variable) value)
       ;;
-    end
 
-    let bindings = { Bindings.opam_version = OpamVersion.to_string OpamVersion.current }
+      let value =
+        let open Values in
+        { opam_version }
+      ;;
+
+      (* Tests that each field matches the constant value for that field. This
+         is included so that [encode] and [decode] can round trip. *)
+      let decode =
+        let open Decoder in
+        let+ loc, opam_version =
+          fields
+          @@ located
+          @@ field Fields.opam_version ~default:Values.opam_version string
+        in
+        if not (String.equal opam_version Values.opam_version)
```

Comment:
Why do we have this check?

Reply:
It's so that users can't encode an env, change a constant field's value, then decode an env with a different value to the contsant's value.

Reply:
Hmm, but I think all this ends up doing is reducing compatibility. Suppose that we update this version in dune because we updated our version of opam. Now we can't build any old build plans even though we really should.

I need to think about this a little more.

Reply:
Ah that's a good point, but I think allowing arbitrary values here is also not the right solution. Do you see a use-case where this data needs to be parsed back into a `Solver_env.t`? I think we're mainly including it here for human-readability. In that case could generate the sovler env in the lockfile with something other than `Solver_env.encode` and not worry about needing to decode it. This would also mean we aren't bound by OSL syntax when serializing the flags field.